Src/IronPython/Runtime/PythonEncoding.cs

```diff
@@ -235,123 +322,179 @@ public abstract class PythonEncoderFallback : EncoderFallback, ICloneable {
-            private ByteCounter _byteCnt;
-            private int _fbkCnt;
+
+            private readonly char _marker;
+
+            private readonly Queue<byte> _allFallbackBytes;  // collects all fallback bytes for the whole pass, only used during actual encoding, pass 2
+            private int _fbkByteCnt; // only used during actual encoding; proxy for _fallbackBytes.Count but valid in pass 1 too
+            private RestorableInt _byteCnt; // counts unreported bytes in the buffer from the last fallback; used during both counting and encoding, but counts separately
+
+            private ReadOnlyMemory<char> _fallbackChars;  // fallback chars (if any) from the last fallback
+            private int _charCnt; // counts unreported chars in the buffer from the last fallback
+            private int _fbkNumChars; // number of all (virtual) chars in the buffer from the last fallback; proxy for _fallbackChars.Length but valid for fallback bytes too
 
             // for error reporting
-            private char _lastCharUnknown;
-            private int _lastIndexUnknown = -1;
+            private int _lastRuneUnknown; // in UTF-32; "rune" is an alias for "Unicode codepoint"
+            private RestorableInt _lastIndexUnknown;
 
             public PythonEncoderFallbackBuffer(bool isPass1, PythonEncoding encoding) {
                 _marker = isPass1 ? Pass1Marker : Pass2Marker;
-                _fallbackBytes = isPass1 ? null : new Queue<byte>();
+                _allFallbackBytes = isPass1 ? null : new Queue<byte>();
                 this.EncodingCharWidth = encoding.CharacterWidth;
                 this.CodePage = encoding.CodePage;
+                _lastIndexUnknown.Reset(-1);
             }
 
+            protected bool EncodingMode { get; private set; }
+
             protected int EncodingCharWidth { get; }
             protected int CodePage { get; }
+            protected char[] Data { get; private set; }
+
+            public virtual void PrepareIncrement(char[] data, bool forEncoding) {
+                Data = data;
+                if (EncodingMode) {
+                    _byteCnt.Save();
+                    _lastIndexUnknown.Save();
+                } else {
+                    _byteCnt.Restore();
+                    _lastIndexUnknown.Restore();
+                }
+                EncodingMode = forEncoding;
+            }
+
+            public abstract Tuple<ReadOnlyMemory<char>, ReadOnlyMemory<byte>> GetFallbackCharsOrBytes(int runeUnknown, int index);
 
-            public abstract byte[] GetFallbackBytes(char charUnknown, int index);
+            public override bool Fallback(char charUnknown, int index)
+                => FallbackImpl(charUnknown, index);
+
+            public override bool Fallback(char charUnknownHigh, char charUnknownLow, int index)
+                => FallbackImpl(char.ConvertToUtf32(charUnknownHigh, charUnknownLow), index);
+
+            private bool FallbackImpl(int runeUnknown, int index) {
```

Comment:
Too bad we can't use `System.Text.Rune`... Maybe we could fake it on .NET Framework?

Reply:
You mean creating this struct but implementing only the used subset of its public members? Or copying over the whole `Rune.cs` from .NET Core and its dependants? Is there a NuGet package available with it?

Reply:
Yes, this is what I meant:
> You mean creating this struct but implementing only the used subset of its public members?

I am not aware of any NuGet package making this available to older runtimes.

Reply:
I think it would be best done in a separate PR: `Rune` is usable in many more places than just `PythonEncoderFallbackBuffer`.

In what namespace would you place this "fake Rune"? `System.Text`? `IronPython.Runtime.Text`? Somewhere else?

Reply:
Probably `IronPython.Runtime.Text` assuming we create such a namespace.